src/Build/Evaluation/PropertyTrackingEvaluatorDataWrapper.cs

```diff
@@ -0,0 +1,230 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Build.BackEnd;
+using Microsoft.Build.Collections;
+using Microsoft.Build.Construction;
+using Microsoft.Build.Execution;
+using Microsoft.Build.Shared.FileSystem;
+using System;
+using System.Collections.Generic;
+using Microsoft.Build.BackEnd.Components.Logging;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Shared;
+using SdkResult = Microsoft.Build.BackEnd.SdkResolution.SdkResult;
+
+namespace Microsoft.Build.Evaluation
+{
+    /// <summary>
+    /// Wraps an existing <see cref="IEvaluatorData{P,I,M,D}"/> allowing the property usage to be tracked.
+    /// </summary>
+    /// <typeparam name="P">The type of properties to be produced.</typeparam>
+    /// <typeparam name="I">The type of items to be produced.</typeparam>
+    /// <typeparam name="M">The type of metadata on those items.</typeparam>
+    /// <typeparam name="D">The type of item definitions to be produced.</typeparam>
+    internal class PropertyTrackingEvaluatorDataWrapper<P, I, M, D> : IEvaluatorData<P, I, M, D>
+        where P : class, IProperty, IEquatable<P>, IValued
+        where I : class, IItem
+        where M : class, IMetadatum
+        where D : class, IItemDefinition<M>
+    {
+        private readonly IEvaluatorData<P, I, M, D> _wrapped;
+        private readonly HashSet<string> _environmentVariables = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+        private readonly EvaluationLoggingContext _evaluationLoggingContext;
+
+        /// <summary>
+        /// Creates an instance of the PropertyTrackingEvaluatorDataWrapper class.
+        /// </summary>
+        /// <param name="dataToWrap">The underlying <see cref="IEvaluatorData{P,I,M,D}"/> to wrap for property tracking.</param>
+        /// <param name="evaluationLoggingContext">The <see cref="EvaluationLoggingContext"/> used to log relevant events.</param>
+        public PropertyTrackingEvaluatorDataWrapper(IEvaluatorData<P, I, M, D> dataToWrap, EvaluationLoggingContext evaluationLoggingContext)
+        {
+            _wrapped = dataToWrap ?? throw new ArgumentNullException(nameof(dataToWrap));
```

Comment:
There's `ErrorUtilities.VerifyThrowInternalNull` and other such helpers for internal assertions.

Reply:
Changed with the latest commit.